torch/csrc/jit/script/compiler.cpp

```diff
@@ -845,15 +747,16 @@ struct to_ir {
                   // never create a Method for the closure
       popFrame(/*ends_def=*/true);
     }
+    AT_ASSERT(block->outputs().size() == 1);
+    // Set the type of the closure equal to its return type.
+    // Since the closures are not-first class yet in the type system,
+    // we use their output type to correctly type forks of closures,
+    // and after the graph is transformed to SSA ensure that forked closures
+    // are actually being invoked on a closure and not an arbitrary value.
+    closure_node->output()->setType(block->outputs().at(0)->type());
```

Comment:
This will make it legal to use a closure as its type, which is just as bad as what was here before. I think it is better to leave closure as a NoneType object and have the fork node look into the definition of the prim::Function node to find the return type.

Reply:
The fork node can't look into the definition of the prim::Function currently because we are emitting the closure as a simple value, so it is stored in the type table, and it is loaded as a `prim::Load`. If we want to be look into the definition i would have to create a Closure Sugared value, which is doable.